python/cffi/source.c

```diff
@@ -0,0 +1,303 @@
+                return 0;
+            }
+        }
+    } else if (node->dflt && !(options & LYP_WD_MASK) && !(node->schema->flags & LYS_CONFIG_R)) {
+        /* LYP_WD_EXPLICIT
+         * - print only if it contains status data in its subtree */
+        LY_TREE_DFS_BEGIN(node, next, elem) {
+            if (elem->schema->flags & LYS_CONFIG_R) {
+                flag = 1;
+                break;
+            }
+            LY_TREE_DFS_END(node, next, elem)
+        }
+        if (!flag) {
+            return 0;
+        }
+    } else if (node->dflt && node->schema->nodetype == LYS_CONTAINER && !(options & LYP_KEEPEMPTYCONT)) {
+        /* avoid empty default containers */
+        LY_TREE_DFS_BEGIN(node, next, elem) {
+            if (elem->schema->nodetype != LYS_CONTAINER) {
+                flag = 1;
+                break;
+            }
+            LY_TREE_DFS_END(node, next, elem)
+        }
+        if (!flag) {
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+/* This function is not exported in the public symbols of libyang.so.
+ * It has been copied here verbatim. */
+int
+lyd_toprint(const struct lyd_node *node, int options)
+{
+    struct lys_node *scase, *sparent;
+    struct lyd_node *first;
+
+    if (!lyd_wd_toprint(node, options)) {
+        /* wd says do not print, but make exception for direct descendants of case nodes without other printable nodes */
+        for (sparent = lys_parent(node->schema); sparent && (sparent->nodetype == LYS_USES); sparent = lys_parent(sparent));
+        if (!sparent || (sparent->nodetype != LYS_CASE)) {
+            /* parent not a case */
+            return 0;
+        }
+        scase = sparent;
+
+        for (sparent = lys_parent(scase); sparent && (sparent->nodetype == LYS_USES); sparent = lys_parent(sparent));
+        if (!sparent || (sparent->nodetype != LYS_CHOICE)) {
+            /* weird */
+            return 0;
+        }
+        if (((struct lys_node_choice *)sparent)->dflt == scase) {
+            /* this is a default case, respect the previous original toprint flag */
+            return 0;
+        }
+
+        /* try to find a sibling that will be printed */
+        for (first = node->prev; first->prev->next; first = first->prev);
+        LY_TREE_FOR(first, first) {
+            if (first == node) {
+                /* skip this node */
+                continue;
+            }
+
+            /* find schema parent, whether it is the same case */
+            for (sparent = lys_parent(first->schema); sparent && (sparent->nodetype == LYS_USES); sparent = lys_parent(sparent));
+            if ((sparent == scase) && lyd_wd_toprint(first, options)) {
+                /* this other node will be printed, we do not have to print the current one */
+                return 0;
+            }
+        }
+
+        /* there is no case child that will be printed, print this node */
+        return 1;
+    }
+
+    return 1;
+}
```

Comment:
As I said, there is some non-API C code that was needed into the python binding. I copied it since I had no way to call it directly. However, it would be best to leave this code into `libyang.so` and make it available in the API somehow.